### PR TITLE
Add painter widget

### PR DESCRIPTION
**Pull request by blink1073**
painter

Interestingly, my canvas only blacks out while licecap is running.

- [x] docs
- [x] example
- [x] tests

### CODE REVIEW

**Comment by tacaswell** [comment]
I think this is going to have to also listen for re-size events to update the offsets if the window is resized.

**Comment by blink1073** [comment]
Done.  I couldn't think of a way to preserve the selection while resizing, so it gets cleared.

File: examples/widgets/painter.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from __future__ import print_function
+"""
+Drag the mouse to paint selected areas of the plot. The callback prints
```
Comment (QuLogic) [inline examples/widgets/painter.py]: Can you add a title and such to match gallery stuff (see [guidelines](https://github.com/matplotlib/matplotlib/blob/master/doc/devel/MEP/MEP12.rst#clean-up-guidelines))?

File: examples/widgets/painter.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+# Generate data and create a scatter plot.
+data = np.random.rand(100, 2)
+data[:, 1] *= 2
+pts = ax.scatter(data[:, 0], data[:, 1], s=80)
+
+
+# Define the "on_select" callback.
+def test(x, y):
```
Comment (QuLogic) [inline examples/widgets/painter.py]: Should have a more specific name.

File: lib/matplotlib/tests/test_widgets.py [hunk 3]
```diff
@@ -271,3 +271,35 @@ def clicked_function():
...
+    tool.radius = 10
+    assert tool.radius == 10
+    tool.label = 2
+    assert tool.label == 2
+
+    do_event(tool, 'press', xdata=100, ydata=100)
+    do_event(tool, 'onmove', xdata=110, ydata=110)
+    do_event(tool, 'release')
+
+    assert tool.overlay[tool.overlay == 2].size == 878
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_widgets.py]: What do these numbers mean?

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
             self.canvas.blit(self.ax.bbox)
         else:
             self.canvas.draw_idle()
+
+
+LABELS_CMAP = ListedColormap(['white', 'red', 'dodgerblue', 'gold',
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Why these colours? Why can't we use an existing one?

File: lib/matplotlib/widgets.py [hunk 5]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+
+        p = Painter(ax, test)
+        plt.show()
+    """
+
+    def __init__(self, ax, on_select=None, overlay_props=None,
+                 cursor_props=None, radius=5, cmap=LABELS_CMAP,
+                 useblit=True, button=None, state_modifier_keys=None):
+        """
+        Parameters:
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Please use numpydoc style for new documentation. Basically all you need is to add some underlining:
```
Parameters
----------
```
and then remove the unnecessary *emphasis* on the parameter names.

File: lib/matplotlib/widgets.py [hunk 6]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+            - The colormap to use for the cursors.
+        *useblit* : bool
+           - Whether to use blitting.
+        *button* : list
+            The button numbers supported for the tool (defaults to [1, 2, 3])
+        *state_modifier_keys* : dict
+            A mapping of key names to state modifiers.
+        """
+        super(Painter, self).__init__(ax, on_select,
+            useblit=useblit, button=button, state_modifier_keys=None)
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: This needs to match the opening `(` to follow PEP8, or there can be no arguments on the first line.

File: lib/matplotlib/widgets.py [hunk 7]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+        Parameters:
+
+        *ax* : :class:`~matplotlib.axes.Axes`
+            The parent axes for the widget
+        *on_select* : function
+           A callback for when a region is painted. Called with the
+           (x, y) coordinates of the region.
+        *overlay_props* : dict
+           The properties to apply to the overlay.
+        *cursor_props* : ditc
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: ditc -> dict

File: lib/matplotlib/widgets.py [hunk 8]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+    """
+
+    def __init__(self, ax, on_select=None, overlay_props=None,
+                 cursor_props=None, radius=5, cmap=LABELS_CMAP,
+                 useblit=True, button=None, state_modifier_keys=None):
+        """
+        Parameters:
+
+        *ax* : :class:`~matplotlib.axes.Axes`
+            The parent axes for the widget
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Missing terminating period.

File: lib/matplotlib/widgets.py [hunk 9]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+        *on_select* : function
+           A callback for when a region is painted. Called with the
+           (x, y) coordinates of the region.
+        *overlay_props* : dict
+           The properties to apply to the overlay.
+        *cursor_props* : ditc
+            - The properties to apply to the cursor.
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Unnecessary bullet point.

File: lib/matplotlib/widgets.py [hunk 10]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+        *radius* : int
+           - The radius of the cursor in pixels.
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Unnecessary bullet point. Also, indent is only 3 spaces instead of 4.

File: lib/matplotlib/widgets.py [hunk 11]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+        *cmap* : :class:~matplotlib.colorls.ListedColormap`
+            - The colormap to use for the cursors.
+        *useblit* : bool
+           - Whether to use blitting.
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Unnecessary bullet point. Also, indent is only 3 spaces instead of 4.

File: lib/matplotlib/widgets.py [hunk 12]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+           A callback for when a region is painted. Called with the
+           (x, y) coordinates of the region.
+        *overlay_props* : dict
+           The properties to apply to the overlay.
+        *cursor_props* : ditc
+            - The properties to apply to the cursor.
+        *radius* : int
+           - The radius of the cursor in pixels.
+        *cmap* : :class:~matplotlib.colorls.ListedColormap`
+            - The colormap to use for the cursors.
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Unnecessary bullet point.

Comment (QuLogic) [inline lib/matplotlib/widgets.py]: How are the alternate colours used? I can't get the example to trigger anything other than a red square.

File: lib/matplotlib/widgets.py [hunk 13]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+    def radius(self):
+        """The label radius in pixels"""
+        return self._radius
+
+    @radius.setter
+    def radius(self, r):
+        self._radius = r
+        xfm = self.ax.transData.inverted()
+        x0, y0 = xfm.transform((0, 0))
+        x1, y1 = xfm.transform((r, r))
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: I'm not totally sure this will work on a polar plot.

File: lib/matplotlib/widgets.py [hunk 14]
```diff
@@ -2595,3 +2598,190 @@ def onmove(self, event):
...
+        self.update()
+        self.onselect(event.xdata, event.ydata)
+
+    def _resize(self, event):
+        extents = self.ax.get_window_extent().extents
+        self._offsetx = extents[0]
+        self._offsety = extents[1]
+        self._shape = (int(extents[3] - extents[1]),
+                       int(extents[2] - extents[0]))
+        self._overlay = np.zeros(self._shape, dtype='uint8')
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: You may be able to use `_image.resample` in the same way [that `imshow` eventually does](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/image.py#L407) if you want take a stab at preserving the overlay.

File: lib/matplotlib/widgets.py (hunk 13 above)
Comment (blink1073) [inline lib/matplotlib/widgets.py]: Added a note that polar plots are not supported (to keep the implementation simpler).

File: lib/matplotlib/widgets.py (hunk 14 above)
Comment (blink1073) [inline lib/matplotlib/widgets.py]: That uses a lot of private APIs, I'd prefer to keep this simple.  I added a note about the clearing behavior in the docs.

**Comment by blink1073** [comment]
Thanks for the review, @QuLogic!  I believe I had addressed all of your concerns.

**Comment by jklymak** [comment]
Is this something we still want for 3.0?  Milestoning as such, but not because I think its a pressing need.  Kinda cool though...

**Comment by blink1073** [comment]
Closing as too far out of date.

**Comment by timhoffm** [comment]
If somebody is interested in this, starting as a 3rd-party-lib is a good way to accumulate user interest and stabilize the API. If both is done, we can reconsider integrating such a widget into core matplotlib later.

**Pull request closed without merging**